### DISCUSSION TITLE
support for ollama

### DISCUSSION THREAD

**Discussion by gtnbssn** [gsuuon/llm.nvim#9] [Q&A]
Does this support [ollama](https://github.com/jmorganca/ollama)?

So far ollama has been my smoothest experience to set up an llm locally.

Here's what the API use looks like (as documented [here](https://ollama.ai/library/phind-codellama)):

```bash
curl -X POST http://localhost:11434/api/generate -d '{
  "model": "phind-codellama",
  "prompt":"Implement a linked list in C++"
 }'
```

**Comment by gsuuon** [comment]
There's currently no ollama support built-in, but this plugin is designed to make adding new providers easy (and can be done externally, as in without modifying llm.nvim). You just need to create a lua table that fulfills [this interface](https://github.com/gsuuon/llm.nvim/blob/f3a3881094389e6ad341fb955e55929c9d1f011a/lua/llm/provider.lua#L15). The [huggingface provider](https://github.com/gsuuon/llm.nvim/blob/f3a3881094389e6ad341fb955e55929c9d1f011a/lua/llm/providers/huggingface.lua) is probably the simplest example. To use the provider, just set it as the provider field in a new prompt, e.g:

```lua
local curl = require('llm.curl')

local ollama = {
  request_completion = function(handlers, params, options)
    return curl.stream(
      -- request goes here
    )
  end
}

require('llm').setup({
  prompts = { -- or vim.tbl_extend('force', require('llm.prompts.starters'),
    ollama_code = {
      provider = ollama,
      -- rest of prompt
    }
  }
})
```
Feel free to open an issue to track ollama support as well - I only have an 8gb MBA so I haven't really played around with local runners which are mac only, but once they add cross platform I'll give it a shot.

**Reply by gsuuon** [reply]
And of course, a PR would be welcome!
